File: android_java/recyclerview_demo/app/src/main/java/com/heyangpeng/recyclerview_demo/MainActivity.java
package com.heyangpeng.recyclerview_demo;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;

import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity implements View.OnClickListener {
    private List<Student> students;
    private RecyclerView recyclerView;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        RecyclerView.LayoutManager layoutManager = new LinearLayoutManager(this);
        recyclerView = findViewById(R.id.recyclerView);
        Button gridView = findViewById(R.id.gridView);
        Button waterfallView = findViewById(R.id.waterfallView);
        gridView.setOnClickListener(this);
        waterfallView.setOnClickListener(this);
        recyclerView.setLayoutManager(layoutManager);
        students = new ArrayList<>();
        initStudents();
    }

    @Override
    protected void onStart() {
        super.onStart();
        ListViewAdapter adapter = new ListViewAdapter(students, this);
        recyclerView.setAdapter(adapter);
    }

    private void initStudents() {
        for (int i = 1; i <= 30; i++) {
            Student student = new Student();
            student.setName("Tom(" + i + ")");
            student.setDesc("Tom(" + i + ")" + "is a good boy!");
            students.add(student);
        }
    }

    @Override
    public void onClick(View v) {
        Intent intent = new Intent();
        switch (v.getId()) {
            case R.id.gridView:
                intent.setClass(this, GridViewActivity.class);
                break;
            case R.id.waterfallView:
                intent.setClass(this, WaterFallViewActivity.class);
                break;
        }
        startActivity(intent);
    }
}

File: android_kotlin/recyclerview_demo/app/src/main/java/com/heyangpeng/recyclerview_demo/MainActivity.kt
package com.heyangpeng.recyclerview_demo

import android.content.Intent
import android.os.Bundle
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity(), View.OnClickListener {
    private lateinit var students: MutableList<Student>

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        val layoutManager: RecyclerView.LayoutManager = LinearLayoutManager(this)
        recyclerView.layoutManager = layoutManager
        gridView.setOnClickListener(this)
        waterfallView.setOnClickListener(this)
        initStudents()
    }

    override fun onStart() {
        super.onStart()
        val listViewAdapter = ListViewAdapter(students)
        recyclerView.adapter = listViewAdapter
    }

    private fun initStudents() {
        students = ArrayList()
        for (i in 1..30) {
            val student = Student("Tom($i)", "Tom($i) is good boy!")
            students.add(student)
        }
    }

    override fun onClick(v: View?) {
        when (v?.id) {
            R.id.gridView -> startActivity(Intent(this, GridViewActivity::class.java))
            R.id.waterfallView -> startActivity(Intent(this, WaterFullViewActivity::class.java))
        }
    }
}
